lib/manager/poetry/extract.ts

```diff
@@ -84,3 +85,20 @@ function extractFromSection(
   });
   return deps;
 }
+
+function extractRegistries(pyprojectfile: PoetryFile): string[] {
+  const sources = pyprojectfile.tool.poetry.source;
+
+  if (!Array.isArray(sources)) {
+    return [];
```

Comment:
Need to check if null or empty array is correct here

Reply:
[`null` isn't a thing in TOML](https://github.com/toml-lang/toml/issues/30). Good point though, I added a test case for empty array and another for missing `source` section.

Reply:
I meant whether we need `repositoryUrls` to be `null` or `[]`.

Reply:
Actually, pypi datasource [drops `pypi.org` registry if a non-empty array is provided](https://github.com/renovatebot/renovate/blob/27fd0f9811d3142f1eb9ff5b839c85c23ca8c3c1/lib/datasource/pypi/index.ts#L37). So I guess it could be `null` or `[]` interchangeably. However for the same reason, if we return a non-empty array of sources, `pypi.org` registry would be dropped which is not the actual behavior of poetry. 93c072d0fa62da60b112a27a6bf23a4af93954eb fixes this.

Reply:
OK, if Poetry always tries pypi.org then we can leave it as-is.